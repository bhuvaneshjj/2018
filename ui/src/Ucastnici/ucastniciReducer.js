export const initialState = { allIds: [], byIds: {} };

const ucastniciReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'RECEIVE_UCASTNICI':
      return action.ucastnici;
    default:
      return state;
  }
};

export default ucastniciReducer;

// Forces 'null' to be the last (desc = false) or first (desc = true).
// Note that 'undefined' is never passed to the sort method (virtue of Array.prototype.sort).
export const narozeniSortMethod = (a, b, desc = false) => {
  if (a && b) {
    if (a.rok < b.rok) {
      return -1;
    }
    if (a.rok > b.rok) {
      return +1;
    }

    if (a.mesic && b.mesic) {
      if (a.mesic < b.mesic) {
        return -1;
      }
      if (a.mesic > b.mesic) {
        return +1;
      }

      if (a.den && b.den) {
        if (a.den < b.den) {
          return -1;
        }
        if (a.den > b.den) {
          return +1;
        }
        return 0; // a tie
      }

      if (!a.den && !b.den) {
        return 0; // both null or undefined
      }
      if (a.den) {
        return desc ? +1 : -1;
      }
      return desc ? -1 : +1;
    }

    if (!a.mesic && !b.mesic) {
      return 0; // both null or undefined
    }
    if (a.mesic) {
      return desc ? +1 : -1;
    }
    return desc ? -1 : +1;
  }

  if (!a && !b) {
    return 0; // both null
  }
  if (a) {
    return desc ? +1 : -1;
  }
  return desc ? -1 : +1;
};

const collator = new Intl.Collator('cs');
export const csStringSortMethod = (a, b) => collator.compare(a, b);

export const getUcastniciOverviewSorted = ({ allIds, byIds }) => {
  const ucastnici = [];
  allIds.forEach(id => {
    const ucastnik = byIds[id];
    const posledniUcast = ucastnik[ucastnik.roky[0]];

    ucastnici.push({
      id,
      prijmeni: posledniUcast.udaje.prijmeni,
      jmeno: posledniUcast.udaje.jmeno,
      narozeni: posledniUcast.udaje.narozeni
    });
  });

  const sorted = ucastnici.sort((a, b) => {
    const prijmeniCmp = csStringSortMethod(a.prijmeni, b.prijmeni);
    if (prijmeniCmp !== 0) {
      return prijmeniCmp;
    }

    const jmenoCmp = csStringSortMethod(a.jmeno, b.jmeno);
    if (jmenoCmp !== 0) {
      return jmenoCmp;
    }

    return narozeniSortMethod(a, b);
  });

  return sorted.map(ucastnik => {
    const { narozeni, ...ostatek } = ucastnik;
    const { den, mesic, rok } = narozeni;
    return { ...ostatek, narozeni: mesic && den ? `${den}. ${mesic}. ${rok}` : `${rok}` };
  });
};
