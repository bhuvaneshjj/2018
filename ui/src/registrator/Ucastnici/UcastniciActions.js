const requestUcastnici = () => ({
  type: 'REQUEST_UCASTNICI'
});

const normalizeUcastnik = json => {
  const roky = [];
  const ucasti = {};

  json.ucasti.forEach(ucast => {
    const { rok, ...ucastBezRoku } = ucast;
    roky.push(rok);
    ucasti[rok] = ucastBezRoku;
  });

  // První element je vždycky nejvyšší rok.
  roky.sort((a, b) => b - a);

  return { roky, ucasti };
};

const normalizeUcastnici = json => {
  const allIds = [];
  const byIds = {};

  json.response.forEach(ucastnik => {
    allIds.push(ucastnik.id);

    const { roky, ucasti } = normalizeUcastnik(ucastnik);
    byIds[ucastnik.id] = { roky, ...ucasti };
  });

  return { allIds, byIds };
};

export const receiveUcastnici = json => ({
  type: 'RECEIVE_UCASTNICI',
  ucastnici: normalizeUcastnici(json),
  receivedAt: Date.now()
});

export const fetchUcastnici = () => async (dispatch, getState, wsClient) => {
  dispatch(requestUcastnici());

  try {
    const response = await wsClient.sendRequest({ action: 'find_all_ucastnici' }); // TODO: from common.findAllUcastnici
    dispatch(receiveUcastnici(response));
  } catch (err) {
    console.log(`Chyba u wsClienta: ${err}`);
    // TODO: dispatch error somehow
  }
};
