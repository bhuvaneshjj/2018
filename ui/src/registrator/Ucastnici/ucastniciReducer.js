export const initialState = { allIds: [], byIds: {} };

const ucastniciReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'RECEIVE_UCASTNICI':
      return action.ucastnici;
    default:
      return state;
  }
};

export default ucastniciReducer;

export const getUcastniciOverviewSorted = ({ allIds, byIds }) => {
  const ucastnici = [];
  allIds.forEach(id => {
    const ucastnik = byIds[id];
    const posledniUcast = ucastnik[ucastnik.roky[0]];

    ucastnici.push({
      id,
      prijmeni: posledniUcast.udaje.prijmeni,
      jmeno: posledniUcast.udaje.jmeno,
      narozeni: posledniUcast.udaje.narozeni.rok
    });
  });

  return ucastnici.sort((a, b) => {
    const prijmeniCmp = a.prijmeni.localeCompare(b, 'cs');
    if (prijmeniCmp !== 0) {
      return prijmeniCmp;
    }

    const jmenoCmp = a.jmeno.localeCompare(b, 'cs');
    if (jmenoCmp !== 0) {
      return jmenoCmp;
    }

    return a.narozeni - b.narozeni;
  });
};
