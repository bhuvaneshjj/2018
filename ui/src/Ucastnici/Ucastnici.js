import React, { Component } from 'react';
import ReactTable from 'react-table';
import PropTypes from 'prop-types';
import 'react-table/react-table.css';
import { csStringSortMethod, narozeniSortMethod } from './ucastniciReducer';
import './Ucastnici.css';

const NarozeniRenderer = row => {
  if (row.value.mesic && row.value.den) {
    return (
      <div>
        {row.value.den}. {row.value.mesic}. {row.value.rok}
      </div>
    );
  }
  return <div>{row.value.rok}</div>;
};

class Ucastnici extends Component {
  componentDidMount = () => {
    const { fetchUcastnici } = this.props;
    fetchUcastnici();
  };

  render = () => {
    const { ucastnici } = this.props;

    if (ucastnici.length === 0) {
      return <div className="Ucastnici">žádný účastník</div>;
    }

    const columns = [
      {
        Header: 'Příjmení',
        accessor: 'prijmeni',
        sortMethod: csStringSortMethod
      },
      {
        Header: 'Jméno',
        accessor: 'jmeno',
        sortMethod: csStringSortMethod
      },
      {
        Header: 'Narození',
        accessor: 'narozeni',
        sortMethod: narozeniSortMethod,
        Cell: NarozeniRenderer
      }
    ];

    return (
      <ReactTable
        className="-striped -highlight"
        data={ucastnici}
        columns={columns}
        showPagination={false}
        defaultPageSize={ucastnici.length}
      />
    );
  };
}

Ucastnici.propTypes = {
  ucastnici: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      prijmeni: PropTypes.string.isRequired,
      jmeno: PropTypes.string.isRequired,
      narozeni: PropTypes.shape({
        rok: PropTypes.number.isRequired,
        mesic: PropTypes.number,
        den: PropTypes.number
      })
    }).isRequired
  ).isRequired,
  fetchUcastnici: PropTypes.func.isRequired
};

export default Ucastnici;
