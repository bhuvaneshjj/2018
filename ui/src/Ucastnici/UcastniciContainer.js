import { connect } from 'react-redux';
import UcastniciResponsive from './UcastniciResponsive';
import { fetchUcastnici } from './UcastniciActions';
import { getUcastniciOverviewSorted } from './ucastniciReducer';

const mapStateToProps = ({ ucastnici }) => ({
  ucastnici: getUcastniciOverviewSorted(ucastnici)
});

const mapDispatchToProps = dispatch => ({
  fetchUcastnici: () => {
    dispatch(fetchUcastnici());
  }
});

const UcastniciContainer = connect(mapStateToProps, mapDispatchToProps)(UcastniciResponsive);

export default UcastniciContainer;
