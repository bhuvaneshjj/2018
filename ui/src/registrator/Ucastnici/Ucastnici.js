import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './Ucastnici.css';

class Ucastnici extends Component {
  componentDidMount = () => {
    const { fetchUcastnici } = this.props;
    fetchUcastnici();
  };

  render = () => {
    const ucastnici = this.props.ucastnici;

    if (ucastnici.length === 0) {
      return <div className="Ucastnici">žádný účastník</div>;
    }

    return (
      <ul className="Ucastnici">
        {ucastnici.map(ucastnik => (
          <li key={ucastnik.id}>
            {ucastnik.prijmeni} {ucastnik.jmeno} {ucastnik.narozeni}
          </li>
        ))}
      </ul>
    );
  };
}

Ucastnici.propTypes = {
  ucastnici: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      prijmeni: PropTypes.string.isRequired,
      jmeno: PropTypes.string.isRequired,
      narozeni: PropTypes.number.isRequired
    }).isRequired
  ).isRequired,
  fetchUcastnici: PropTypes.func.isRequired
};

export default Ucastnici;
