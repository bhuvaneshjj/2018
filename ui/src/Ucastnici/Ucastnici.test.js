import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Ucastnici from './Ucastnici';

const fakeFetchUcastnici = () => {};

it('žádný účastník', () => {
  const wrapper = shallow(
    <Ucastnici
      ucastnici={[]}
      fetchUcastnici={fakeFetchUcastnici}
      containerWidth={500}
      containerHeight={500}
    />
  );
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('dva účastníci', () => {
  const ucastnici = [
    {
      id: '5a09b1fd371dec1e99b7e1c9',
      prijmeni: 'Balabák',
      jmeno: 'Roman',
      narozeni: '1956'
    },
    {
      id: '6f09b1fd371dec1e99b7e1c9',
      prijmeni: 'Sukdoláková',
      jmeno: 'Martina',
      narozeni: '7. 12. 1963'
    }
  ];

  const wrapper = shallow(
    <Ucastnici
      ucastnici={ucastnici}
      fetchUcastnici={fakeFetchUcastnici}
      containerWidth={500}
      containerHeight={500}
    />
  );
  expect(toJSON(wrapper)).toMatchSnapshot();
});
