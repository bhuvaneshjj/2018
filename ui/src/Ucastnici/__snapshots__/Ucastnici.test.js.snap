// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dva účastníci 1`] = `
<ReactTable
  AggregatedComponent={[Function]}
  ExpanderComponent={[Function]}
  FilterComponent={[Function]}
  LoadingComponent={[Function]}
  NoDataComponent={[Function]}
  PadRowComponent={[Function]}
  PaginationComponent={[Function]}
  PivotValueComponent={[Function]}
  ResizerComponent={[Function]}
  TableComponent={[Function]}
  TbodyComponent={[Function]}
  TdComponent={[Function]}
  TfootComponent={[Function]}
  ThComponent={[Function]}
  TheadComponent={[Function]}
  TrComponent={[Function]}
  TrGroupComponent={[Function]}
  aggregatedKey="_aggregated"
  className="-striped -highlight"
  collapseOnDataChange={true}
  collapseOnPageChange={true}
  collapseOnSortingChange={true}
  column={
    Object {
      "Aggregated": undefined,
      "Cell": undefined,
      "Expander": undefined,
      "Filter": undefined,
      "Footer": undefined,
      "Header": undefined,
      "Pivot": undefined,
      "PivotValue": undefined,
      "aggregate": undefined,
      "className": "",
      "filterAll": false,
      "filterMethod": undefined,
      "filterable": undefined,
      "footerClassName": "",
      "footerStyle": Object {},
      "getFooterProps": [Function],
      "getHeaderProps": [Function],
      "getProps": [Function],
      "headerClassName": "",
      "headerStyle": Object {},
      "minWidth": 100,
      "resizable": undefined,
      "show": true,
      "sortMethod": undefined,
      "sortable": undefined,
      "style": Object {},
    }
  }
  columns={
    Array [
      Object {
        "Header": "Příjmení",
        "accessor": "prijmeni",
        "sortMethod": [Function],
      },
      Object {
        "Header": "Jméno",
        "accessor": "jmeno",
        "sortMethod": [Function],
      },
      Object {
        "Cell": [Function],
        "Header": "Narození",
        "accessor": "narozeni",
        "sortMethod": [Function],
      },
    ]
  }
  data={
    Array [
      Object {
        "id": "5a09b1fd371dec1e99b7e1c9",
        "jmeno": "Roman",
        "narozeni": Object {
          "rok": 1956,
        },
        "prijmeni": "Balabák",
      },
      Object {
        "id": "6f09b1fd371dec1e99b7e1c9",
        "jmeno": "Martina",
        "narozeni": Object {
          "den": 7,
          "mesic": 12,
          "rok": 1963,
        },
        "prijmeni": "Sukdoláková",
      },
    ]
  }
  defaultExpanded={Object {}}
  defaultFilterMethod={[Function]}
  defaultFiltered={Array []}
  defaultPageSize={2}
  defaultResized={Array []}
  defaultSortDesc={false}
  defaultSortMethod={[Function]}
  defaultSorted={Array []}
  expanderDefaults={
    Object {
      "filterable": false,
      "resizable": false,
      "sortable": false,
      "width": 35,
    }
  }
  filterable={false}
  freezeWhenExpanded={false}
  getLoadingProps={[Function]}
  getNoDataProps={[Function]}
  getPaginationProps={[Function]}
  getProps={[Function]}
  getResizerProps={[Function]}
  getTableProps={[Function]}
  getTbodyProps={[Function]}
  getTdProps={[Function]}
  getTfootProps={[Function]}
  getTfootTdProps={[Function]}
  getTfootTrProps={[Function]}
  getTheadFilterProps={[Function]}
  getTheadFilterThProps={[Function]}
  getTheadFilterTrProps={[Function]}
  getTheadGroupProps={[Function]}
  getTheadGroupThProps={[Function]}
  getTheadGroupTrProps={[Function]}
  getTheadProps={[Function]}
  getTheadThProps={[Function]}
  getTheadTrProps={[Function]}
  getTrGroupProps={[Function]}
  getTrProps={[Function]}
  groupedByPivotKey="_groupedByPivot"
  indexKey="_index"
  loading={false}
  loadingText="Loading..."
  nestingLevelKey="_nestingLevel"
  nextText="Next"
  noDataText="No rows found"
  ofText="of"
  onFetchData={[Function]}
  originalKey="_original"
  pageSizeOptions={
    Array [
      5,
      10,
      20,
      25,
      50,
      100,
    ]
  }
  pageText="Page"
  pivotDefaults={Object {}}
  pivotIDKey="_pivotID"
  pivotValKey="_pivotVal"
  previousText="Previous"
  resizable={true}
  rowsText="rows"
  showPageJump={true}
  showPageSizeOptions={true}
  showPagination={false}
  showPaginationBottom={true}
  showPaginationTop={false}
  sortable={true}
  style={Object {}}
  subRowsKey="_subRows"
/>
`;

exports[`žádný účastník 1`] = `
<div
  className="Ucastnici"
>
  žádný účastník
</div>
`;
